src/BackupChain.cs

```diff
@@ -54,51 +54,51 @@ private BackupChain(IList<BackupMetadata> recentBackups)
-    {
-      var diffBackupsOrdered = backups
-        .Where(b =>
-                 b.BackupType == BackupFileTools.BackupType.Diff &&
-                 b.DatabaseBackupLsn == lastFullBackup.CheckpointLsn)
-        .OrderByDescending(b => b.LastLsn).ToList();
-
-      if(!diffBackupsOrdered.Any())
-        return new List<BackupMetadata>();
-
-      var targetLastLsn = diffBackupsOrdered.First().LastLsn;
-      // get all the stripes of this backup
-      return diffBackupsOrdered.Where(diffBackup => diffBackup.LastLsn == targetLastLsn);
-    }
-
-    private static IEnumerable<BackupMetadata> NextLogBackup(IEnumerable<BackupMetadata> backups,
-      BackupMetadata prevBackup)
-    {
-      // also gets all the stripes of the next backup
-      return backups.Where(b => b.BackupType == BackupFileTools.BackupType.Log &&
-                                prevBackup.LastLsn >= b.FirstLsn && prevBackup.LastLsn + 1 < b.LastLsn);
-    }
-
-    private static bool IsValidFilePath(BackupMetadata meta)
-    {
-      var path = meta.PhysicalDeviceName;
-      return BackupFileTools.IsValidFileUrl(path) || BackupFileTools.IsValidFilePath(path);
-    }
-  }
-}
\ No newline at end of file
+    public IEnumerable<BackupMetadata> OrderedBackups => _orderedBackups;
+
+    private static IEnumerable<BackupMetadata> MostRecentFullBackup(IEnumerable<BackupMetadata> backups)
+    {
+      var fullBackupsOrdered = backups
+        .Where(b => b.BackupType == BackupFileTools.BackupType.Full)
+        .OrderByDescending(d => d.CheckpointLsn).ToList();
+
+      if(!fullBackupsOrdered.Any())
+        throw new BackupChainException("Could not find any full backups");
+
+      var targetCheckpointLsn = fullBackupsOrdered.First().CheckpointLsn;
+      // get all the stripes of this backup
+      return fullBackupsOrdered.Where(fullBackup => fullBackup.CheckpointLsn == targetCheckpointLsn);
+    }
+
+    private static IEnumerable<BackupMetadata> MostRecentDiffBackup(IEnumerable<BackupMetadata> backups,
+      BackupMetadata lastFullBackup)
+    {
+      var diffBackupsOrdered = backups
+        .Where(b =>
+                 b.BackupType == BackupFileTools.BackupType.Diff &&
+                 b.DatabaseBackupLsn == lastFullBackup.CheckpointLsn)
+        .OrderByDescending(b => b.LastLsn).ToList();
+
+      if(!diffBackupsOrdered.Any())
+        return new List<BackupMetadata>();
+
+      var targetLastLsn = diffBackupsOrdered.First().LastLsn;
+      // get all the stripes of this backup
+      return diffBackupsOrdered.Where(diffBackup => diffBackup.LastLsn == targetLastLsn);
+    }
+
+    private static IEnumerable<BackupMetadata> NextLogBackup(IEnumerable<BackupMetadata> backups,
+      BackupMetadata prevBackup)
+    {
+      // also gets all the stripes of the next backup
+      return backups.Where(b => b.BackupType == BackupFileTools.BackupType.Log &&
+                                prevBackup.LastLsn >= b.FirstLsn && prevBackup.LastLsn <= b.LastLsn);

```

Comment:
I'm wondering, if this is a simpler check?
```csharp
prevBackup.FirstLsn <= b.FirstLsn && prevBackup.LastLsn <= b.LastLsn
```

Reply:
no, the point of chaining is to check first_lsn vs last_lsn